metrics/internal/cache/rbd-mirror.go

```diff
@@ -100,21 +121,70 @@ var _ cache.Store = &RBDMirrorStore{}
 type RBDMirrorStore struct {
 	Mutex sync.RWMutex
 	// Store is a map of Pool UID to RBDMirrorPoolStatusVerbose
-	Store           map[types.UID]RBDMirrorPoolStatusVerbose
-	rbdCommandInput rbdCommandInput
+	Store map[types.UID]RBDMirrorPoolStatusVerbose
+	// rbdCommandInput is a struct that contains the input for the rbd command
+	// for each AllowdNamespaces
+	rbdCommandInput   map[string]*rbdCommandInput
+	kubeclient        clientset.Interface
+	allowedNamespaces []string
 }
 
-func NewRBDMirrorStore() *RBDMirrorStore {
+func NewRBDMirrorStore(opts *options.Options) *RBDMirrorStore {
+	// write Ceph config file before issuing RBD mirror commands
+	err := writeCephConfig()
+	if err != nil {
+		// With the current implementation, this is not possible.
+		panic(err)
+	}
 	return &RBDMirrorStore{
-		Store:           map[types.UID]RBDMirrorPoolStatusVerbose{},
-		rbdCommandInput: rbdCommandInput{},
+		Store:             map[types.UID]RBDMirrorPoolStatusVerbose{},
+		rbdCommandInput:   map[string]*rbdCommandInput{},
+		kubeclient:        clientset.NewForConfigOrDie(opts.Kubeconfig),
+		allowedNamespaces: opts.AllowedNamespaces,
 	}
 }
 
-func (s *RBDMirrorStore) WithRBDCommandInput(monitor, id, key string) {
-	s.rbdCommandInput.monitor = monitor
-	s.rbdCommandInput.id = id
-	s.rbdCommandInput.key = key
+func (s *RBDMirrorStore) WithRBDCommandInput(namespace string) error {
+	var allow bool
+	for _, item := range s.allowedNamespaces {
+		if item == namespace {
+			allow = true
+			break
+		}
+	}
+	if !allow {
+		return fmt.Errorf("rbd-mirror metrics collection from namespace %q is not allowed", namespace)
+	}
+
+	secret, err := s.kubeclient.CoreV1().Secrets(namespace).Get(context.TODO(), "rook-ceph-mon", v1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get secret in namespace %q: %v", namespace, err)
+	}
+	key, ok := secret.Data["ceph-secret"]
+	if !ok {
+		return fmt.Errorf("failed to get client key from secret in namespace %q", namespace)
+	}
+	id := "admin"
+
+	configmap, err := s.kubeclient.CoreV1().ConfigMaps(namespace).Get(context.TODO(), "rook-ceph-csi-config", v1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get configmap in namespace %q: %v", namespace, err)
+	}
+
+	data := configmap.Data["csi-cluster-config-json"]
+	var clusterConfig []csiClusterConfig
+	err = json.Unmarshal([]byte(data), &clusterConfig)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal csi-cluster-config-json in namespace %q: %v", namespace, err)
+	}
+
+	input := rbdCommandInput{}
+	input.monitor = clusterConfig[0].Monitors[0]
```

Comment:
this might panic if no mons exits or no clusterConfig exists?

Reply:
true. added some checks to handle such cases.